pkg/ingest/partition_offset_client.go

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingest
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/kmsg"
+)
+
+const (
+	// kafkaOffsetStart is a special offset value that means the beginning of the partition.
+	kafkaOffsetStart = int64(-2)
+
+	// kafkaOffsetEnd is a special offset value that means the end of the partition.
+	kafkaOffsetEnd = int64(-1)
+
+	// ReaderMetricsPrefix is the reader metrics prefix used by the ingest storage.
+	ReaderMetricsPrefix = "cortex_ingest_storage_reader"
+)
+
+// partitionOffsetClient is a client used to read partition offsets.
+type PartitionOffsetClient struct {
+	client *kgo.Client
+	admin  *kadm.Client
+
+	topic string
+}
+
+func NewPartitionOffsetClient(client *kgo.Client, topic string) *PartitionOffsetClient {
+	return &PartitionOffsetClient{
+		client: client,
+		admin:  kadm.NewClient(client),
+		topic:  topic,
+	}
+}
+
+// FetchPartitionsLastProducedOffsets fetches and returns the last produced offsets for all input partitions. The offset is
+// -1 if a partition has been created but no record has been produced yet. The returned offsets for each partition
+// are guaranteed to be always updated (no stale or cached offsets returned).
+//
+// The Kafka client used under the hood may retry a failed request until the retry timeout is hit.
+func (p *PartitionOffsetClient) FetchPartitionsLastProducedOffsets(ctx context.Context, partitionIDs []int32) (_ kadm.ListedOffsets, returnErr error) {
+	// Skip lookup and don't track any metric if no partition was requested.
+	if len(partitionIDs) == 0 {
+		return nil, nil
+	}
+	return p.fetchPartitionsOffsets(ctx, kafkaOffsetEnd, partitionIDs)
+}
+
+func (p *PartitionOffsetClient) FetchPartitionsStartProducedOffsets(ctx context.Context, partitionIDs []int32) (_ kadm.ListedOffsets, returnErr error) {
+	// Skip lookup and don't track any metric if no partition was requested.
+	if len(partitionIDs) == 0 {
+		return nil, nil
+	}
+
+	return p.fetchPartitionsOffsets(ctx, kafkaOffsetStart, partitionIDs)
+}
+
+// fetchPartitionsEndOffsets returns the end offset of each partition in input. This function returns
+// error if fails to get the offset of any partition (no partial result is returned).
+func (p *PartitionOffsetClient) fetchPartitionsOffsets(ctx context.Context, timestamp int64, partitionIDs []int32) (kadm.ListedOffsets, error) {
```

Comment:
The mix of offset and timestamp is a bit confusing. It would be nice if we could chose one or link to where this is documented (https://github.com/twmb/franz-go/blob/master/pkg/kmsg/generated.go#L5781-L5808)